pysma/const.py

```diff
@@ -0,0 +1,826 @@
+"""Constants for SMA WebConnect library for Python. """
+URL_LOGIN = "/dyn/login.json"
+URL_LOGOUT = "/dyn/logout.json"
+URL_VALUES = "/dyn/getValues.json"
+URL_LOGGER = "/dyn/getLogger.json"
+URL_DASH_LOGGER = "/dyn/getDashLogger.json"
+URL_DASH_VALUES = "/dyn/getDashValues.json"
+
+USERS = {"user": "usr", "installer": "istl"}
+
+JMESPATH_BASE = "result.*"
+
+JMESPATH_VAL = "val"
+JMESPATH_VAL_TAG = JMESPATH_VAL + "[0].tag"
+JMESPATH_VAL_IDX = '"{}"[{}].val'
+JMESPATH_VAL_IDX_TAG = JMESPATH_VAL_IDX + "[0].tag"
+JMESPATHS_TAG = (JMESPATH_VAL_IDX_TAG, JMESPATH_VAL_TAG)
+
+LEGACY_MAP = {
+    "pv_power": {"old_key": "6100_0046C200", "new_sensor": "pv_power_a"},
+    "pv_voltage": {
+        "old_key": "6380_40451F00",
+        "new_sensor": "pv_power_a",
+    },
+    "pv_current": {
+        "old_key": "6380_40452100",
+        "new_sensor": "pv_current_a",
+    },
+}
+
+FALLBACK_DEVICE_INFO = {
+    "manufacturer": "SMA",
+    "name": "SMA Device",
+    "type": "",
+    "serial": "9999999999",
+}
+
+DEVCLASS_INVERTER = "1"
+DEVCLASS_BATTERY = "7"
+DEVCLASS_ENERGY_METER = "65"
+OPTIMIZERS_VIA_INVERTER = "253"
+ENERGY_METER_VIA_INVERTER = "254"
+DEVICE_INFO = "255"
+
+SNAME = "name"
+SKEY = "key"
+SUNIT = "unit"
+SFACTOR = "factor"
+SPATH = "path"
+SL10N = "l10n_translate"
+
+# Status - Operation
+SENSOR_STATUS = {
```

Comment:
I do not like these sensors definitions at all. Alternatives could be: nametuple, data classes or even attrs (as used in the original). I used attrs because it was before data classes, but even today it is efficient with `__slots__`.

So another proposal:
- create a `sensors.py`
- move all these sensor definitions there (closer to the old style)
- move the existing @attrs Sensor class to this `sensors.py` files (not to have circular imports)
- potentially leave the methods in `__init__.py` (`extract_value`, `extract_logger`) and refactor them to accept a Sensor as first arg instead of self.

Then you end up with a much more compact sensors file, with the class definition & all sensors. And a sensor is 1 line again instead of 5+, which is difficult to read.

One step further, you could add a method to the Sensor class to push it to the devclass array, so you can have the following, and cut out all these global vars. The result should be very short & declaritive:

```
Sensor("6100_00465700", "frequency", "Hz", 100).used_by(ENERGY_METER, BATTERY)

SENSOR_MAP={}

class Sensor....
  def used_by(*args):
    for arg in args:
      SENSOR_MAP.setdefault(arg, []).append(arg)
```

Reply:
The original `Sensor` class is still used, but now only the sensors that are actually needed are created as object. (https://github.com/rklomp/pysma/blob/e7357369d9aaeb2c60e37e22bb04b93157c1d111/pysma/__init__.py#L170-L172)

If you see the `const.py` as a "config" file for the sensors, the current structure actually makes sense. It also makes it easy for other users to propose changes without having to write any code besides defining dicts. Although I do agree that data structure wise it might not be the cleanest solution and I am open to suggestions to improve it.


Some considerations:
1. Sensors should easily be referenced.
   - For example in `get_devclass` I want to query a status sensor only. https://github.com/rklomp/pysma/blob/e7357369d9aaeb2c60e37e22bb04b93157c1d111/pysma/__init__.py#L381
   - In `get_sensors` I use two sensors to query information to determine which sensors to return to the user. https://github.com/rklomp/pysma/blob/e7357369d9aaeb2c60e37e22bb04b93157c1d111/pysma/__init__.py#L409-L410
2. Sensors should be able to be a "template"
   - For optimizers I create sensors on the fly based on the number of optimzers detected https://github.com/rklomp/pysma/blob/e7357369d9aaeb2c60e37e22bb04b93157c1d111/pysma/__init__.py#L445-L449

---
A different approach might be:
- Have one instance of `Sensors` that contains all sensor definitions (like it was before)
- Add a new `used_by` parameter to `Sensor` that lists all associations.
- Check the `Sensor.used_by` in a for loop in `SMA.get_sensors()` and construct a new `Sensors` object to return to the user.
  - Or add a new `filter(used_by: Enum)` method to `Sensors` that limits the returned sensors by `__iter__` to the sensors matching `used_by`.

This will make the sensors more easily referenced (point 1 above) as you would be able to retreive any Sensor using `Sensors.__getitem__`.

Point 2, templating might be accomplishes by creating a copy of the `Sensor`, updating `key_idx` and `name`. And adding that copy to the newly constructed `Sensors` object.


```python
Sensor("6100_00465700", "frequency", "Hz", 100).used_by(DEVCLASS_INVERTER, DEVCLASS_BATTERY)

class Sensor:
  _used_by = attr.ib(default=[])

  def used_by(self, *args):
    for arg in args:
      self._used_by.append(arg)

  def is_used_by(self, arg):
    return arg in self._used_by

# Assuming self.sensor_map is a Sensors object containing all sensors
class SMA:
  async def get_sensors(self):
    devclass = await self.get_devclass() or DEVCLASS_INVERTER
    device_sensors = [sensor in self.sensor_map if sensor.is_used_by(devclass)]
```

Reply:
> RK> The original Sensor class is still used, but now only the sensors that are actually needed are created as object.

Understand, but now all sensors are dictionary objects, with "loose" key references, using the consts for keys... losing any static type checking advantages in the process. And then for the ones you used you create additional Sensor objects.

This seems like the ideal reason why things like `nametuple` & data classes exist,- hence the proposal to still keep a mostly const like (moving methods from the actual Sensor class and just keeping the barebaone attributes)

With 100 other examples visible at the same time on a screen, even non-coders will be able to duplicate & extend a single line.

> RK> For example in get_devclass I want to query a status sensor only. 

And where this is required (limited mostly to status) there can still be a const SENSOR_STATUS or the likes

> RK> Sensors should be able to be a "template"

And namedtuples/dataclasses etc can still be used to achieve exactly this, using elegant attributes to retrive values and not [MYCONSTKEY]

--- 

your alternative approach could work, I hv eno objection there

Since you already have other attrs, this should also work
```
Sensor("6100_00465700", "frequency", "Hz", 100, used_by=(DEVCLASS_INVERTER, DEVCLASS_BATTERY))

class Sensor:
  ... snip...
  used_by = attr.ib(default=[])
```

Reply:
I've regexed all sensors to a new dataclass format: 14f037f
This needs more work to get it working, but at least shows how this will look like.

Reply:
I have now assigned all `Sensor` objects to a variable and still use the `sensor_map`. While we could use a `used_by` method and programmatically construct this map as you proposed, we could also keep it like this. It gives a good overview of the sensors assigned to each device class and since it is static the `used_by` method does not have any added value as far as I can see.

Also separate sensors can easily be referenced by name. For example: `pysma.sensors.status`

Let me know what you think
I agree that this way it already looks a lot cleaner and is better readable.

Reply:
Indeed, this looks much cleaner!

btw, did you change pylint to allow lower case constants? (no objections, just missed that ignore)

I would still prefer using using `__slots__`, motivation [here](https://github.com/ericvsmith/dataclasses/issues/28#issuecomment-535928781) 
...and since we already have *attrs* as a dependency, maybe just use that instead of the stdlib's data class?

Reply:
> btw, did you change pylint to allow lower case constants? (no objections, just missed that ignore)

pylint should use default settings. I did not add that ignore.

> I would still prefer using using `__slots__`, motivation [here](https://github.com/ericvsmith/dataclasses/issues/28#issuecomment-535928781)
> ...and since we already have _attrs_ as a dependency, maybe just use that instead of the stdlib's data class?

We could also use `NamedTuple` as that also uses slots.

---
What do you think about wrapping the lists in de sensor_map in a call to `Sensors`. This will add the extra functions provided by `Sensors` and there is no need to do that in the `get_sensors` anymore. We could add an `__add__` method to easily merge two `Sensors` objects


```python
sensor_map = {
    DEVCLASS_INVERTER: Sensors(
        [
            status,
            pv_power_a,
            pv_power_b,
...
```

Reply:
You mean [typing.NamedTuple](https://docs.python.org/3/library/typing.html#typing.NamedTuple)?

Referring to the `Sensors` class in `sensors.py` makes it much less of a nice clean constants file, which is is now. The only imports are stdlib (and maybe attrs 😉 ) so anyone can understand it quickly and add sensors

Reply:
also no need to instantiate things you most probably wont use, just keep it a constant array (or even tuple)

Reply:
Never used that typing.NamedTuple, but looks like it could fit the bill for this kind of constant data

Reply:
I have moved `Sensors` to `sensors.py` as I think that is the most logical place.
`Sensor` is now back to attr and `extract_value` moved to `SMA` 96c51b9a48ff62f288b931ce80a40c6e6dbb8ab0

Will continue later this week to get everything fixed up and passing all tests again.

Reply:
Maye you can even have a `__init__(self, *sensors):` constructor on sensors, and skip passing an array in your device-sensor maps

Reply:
I just realize this is not going to work properly for querying multiple devices.
When you create two instances of `SMA` and request a list of sensors, both lists will actually point to the same sensors. The result is that they both try to write their value to the same object.

In your original version this was not an issues because they were created for every newly created `Sensors`

I think it makes sense to only initialize the `Sensor` objects when needed. With my `dict` solution this was not an issues since the dict was fed as arguments to a new `Sensor` object when adding it to `Sensors`. https://github.com/rklomp/pysma/blob/e7357369d9aaeb2c60e37e22bb04b93157c1d111/pysma/__init__.py#L170-L172

A solution would be to have the `add` method of `Sensors` add a copy of the object to the lists, but its not that efficient to have them all created first in `sensors.py` and then copied again when actually needed.

Another way is to create a `NamedTuple` as "template" for example:
```python 
SensorDefinition = namedtuple('SensorDefinition', 'name key unit factor path enabled l10n_translate')

pv_power_a = SensorDefinition("6380_40251E00_0", "pv_power_a", unit="W")
```

And when this is passed to the `add` method of `Sensors`  it will take its parameters and create a new `Sensor` object.
`namedtuple._asdict()` would basically be the `dict` based file I initially created.
```python
        if isinstance(sensor, SensorDefinition):
            self.add(Sensor(**sensor._asdict()))
            return
```

Do you have any other/better ideas?

Reply:
Or move all senor definitions to a `yaml` format and load that from the python code. Basically doing the same as my initial `dict` format, but in a more readable format.

```yaml
sensors:
  - pv_power_a:
      key: 6380_40251E00_0
      unit: W
```
Or as onliner
```yaml
sensors:
  - pv_power_a: {key: 6380_40251E00_0, unit: W}
```

Reply:
I've implemented the `SenorDefinition` namedtuple
https://github.com/rklomp/pysma/compare/e735736...c433630

But that made me have two almost similar classes. I converted the named tuple to a attr based class and the exitsting `Sensor` class is now extending the `SensorDefinition`. See 35df98b

Let me know what you think :)

Reply:
A single `Sensor` class might make it a bit less complex and I dont think the `SensorDefinition` has that much added value.
Let me know which version you prefer. d10bd2a vs 2d0d618

Reply:
The single `Sensor` seems to be the way to go!

Reply:
> Or move all senor definitions to a yaml format and load that from the python code. Basically doing the same as my initial dict format, but in a more readable format.

Sorry, I missed this.

The single `Sensor` class seems good, it also offers some static code checking

Reply:
Great! If you could give it one more review I think this one is ready to be merged 😃